### PR TITLE
CTCP-182: Submit a Departure Declaration XML

### PR DESCRIPTION
**Pull request by liandra-giacomine**
Route departure POST request according to accept header

### CODE REVIEW

File: app/controllers/DeparturesController.scala [hunk 1]
```diff
@@ -62,13 +60,17 @@ class DeparturesController @Inject() (
...
+    with VersionSwitch {
 
   import MetricsKeys.Endpoints._
 
   lazy val departuresCount = histo(GetDeparturesForEoriCount)
 
   def submitDeclaration(): Action[NodeSeq] =
+    versionSwitch(submitDeclarationVersionOne, submitDeclarationVersionTwo)
+
+  def submitDeclarationVersionOne(): Action[NodeSeq] =
```
Comment (derekspalmer) [inline app/controllers/DeparturesController.scala:73]: private method?

Comment (derekspalmer) [reply app/controllers/DeparturesController.scala:73]: same for versionTwo.

File: app/controllers/VersionSwitch.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    Action.async(parse.xml) {
+    (request: Request[NodeSeq]) =>
+      request.headers.get("accept") match {
+        case Some("application/vnd.hmrc.2.0+json") => v2Action(request)
+
+        case Some("application/vnd.hmrc.1.0+json") => v1Action(request)
+
+        case None => v1Action(request)
+
+        case Some(headerVal) => Future.successful(UnsupportedMediaType(s"Unsupported Accept-header: $headerVal"))
```
Comment (derekspalmer) [inline app/controllers/VersionSwitch.scala:38]: Is it worth adding logging here?

**Review by derekspalmer — APPROVED** [review]
LGTM

File: app/controllers/VersionSwitch.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import scala.concurrent.Future
+import scala.xml.NodeSeq
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+trait VersionSwitch {
+ self: BackendController =>
+
+  def versionSwitch(v1Action: Action[NodeSeq], v2Action: Action[NodeSeq]): Action[NodeSeq] =
+    Action.async(parse.xml) {
+    (request: Request[NodeSeq]) =>
```
Comment (derekspalmer) [inline app/controllers/VersionSwitch.scala:30]: type superfluous.

Comment (drnaylor-oc) [reply app/controllers/VersionSwitch.scala:30]: It's not necessarily. When originally building this, we were having type inference issues causing the compiler to pick the wrong overload to `Action#apply`, as a `Request[_]` is also a `RequestHeader`. I haven't checked it here, but I'd imagine that it's possible the same thing is happening here.

This is a case where I think I'd prefer to keep it to ensure the compiler does the right thing -- even if removing it now works for some reason.

Comment (derekspalmer) [reply app/controllers/VersionSwitch.scala:30]: Not sure about that - if the inferencer can't determine the type, it will let us know!
But ok - leave it.

Comment (drnaylor-oc) [reply app/controllers/VersionSwitch.scala:30]: In this case, it wasn't that it couldn't infer a type, it's that it infered a super type of what it actually was - we needed to correct that.

**Review by drnaylor-oc — APPROVED** [review]

**Review by drnaylor-oc — APPROVED** [review]

**Pull request merged**